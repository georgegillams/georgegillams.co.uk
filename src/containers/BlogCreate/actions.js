import { createRoutine } from 'redux-saga-routines';

import { CREATE_BLOG } from './constants';

export const createBlog = createRoutine(CREATE_BLOG);
