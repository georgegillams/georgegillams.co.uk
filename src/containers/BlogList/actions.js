import { createRoutine } from 'redux-saga-routines';

import { LOAD_BLOGS, DELETE_BLOG } from './constants';

export const loadBlogs = createRoutine(LOAD_BLOGS);
export const deleteBlog = createRoutine(DELETE_BLOG);
