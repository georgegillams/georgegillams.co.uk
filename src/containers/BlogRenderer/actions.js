import { createRoutine } from 'redux-saga-routines';

import { LOAD_BLOG, UPDATE_BLOG } from './constants';

export const loadBlog = createRoutine(LOAD_BLOG);
export const updateBlog = createRoutine(UPDATE_BLOG);
