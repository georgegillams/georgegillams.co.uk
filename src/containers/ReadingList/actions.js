import { createRoutine } from 'redux-saga-routines';

import { CREATE_BOOK, LOAD_BOOKS, UPDATE_BOOK, DELETE_BOOK, SET_HIGHLIGHTED_BOOK } from './constants';

export const loadBooks = createRoutine(LOAD_BOOKS);
export const createBook = createRoutine(CREATE_BOOK);
export const updateBook = createRoutine(UPDATE_BOOK);
export const deleteBook = createRoutine(DELETE_BOOK);
export const setHighlightedBook = createRoutine(SET_HIGHLIGHTED_BOOK);
