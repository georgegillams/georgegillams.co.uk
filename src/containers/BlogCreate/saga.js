import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';
import { selectState } from './selectors';

import { createBlog } from './actions';

export function* doCreateBlog() {
  const currentState = yield select(selectState());
  const { blogToCreate } = currentState;
  const requestURL = apiStructure.createBlog.fullPath;

  try {
    yield put(createBlog.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(blogToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(createBlog.failure(result));
    } else {
      yield put(createBlog.success(result));
      window.location = `/blog/${result.id}`;
    }
  } catch (err) {
    yield put(createBlog.failure(err));
  } finally {
    yield put(createBlog.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(createBlog.TRIGGER, doCreateBlog);
}
