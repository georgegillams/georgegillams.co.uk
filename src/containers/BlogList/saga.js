import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'utils/common/request';
import apiStructure from 'helpers/apiStructure';
import { selectState } from './selectors';

import { loadBlogs, deleteBlog } from './actions';

export function* doLoadBlogs() {
  const requestURL = apiStructure.loadBlogs.fullPath;

  try {
    yield put(loadBlogs.request());

    const result = yield call(request, requestURL, {
      method: 'GET',
    });

    if (result.error) {
      yield put(loadBlogs.failure(result));
    } else {
      yield put(loadBlogs.success(result));
    }
  } catch (err) {
    yield put(loadBlogs.failure(err));
  } finally {
    yield put(loadBlogs.fulfill());
  }
}

export function* doDeleteBlog() {
  const currentState = yield select(selectState());
  const { blogToDelete } = currentState;
  const requestURL = apiStructure.deleteBlog.fullPath;

  try {
    yield put(deleteBlog.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ id: blogToDelete }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(deleteBlog.failure(result));
    } else {
      yield put(deleteBlog.success());
    }
    yield put(loadBlogs.trigger());
  } catch (err) {
    yield put(deleteBlog.failure(err));
  } finally {
    yield put(deleteBlog.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadBlogs.TRIGGER, doLoadBlogs);
  yield takeLatest(deleteBlog.TRIGGER, doDeleteBlog);
}
