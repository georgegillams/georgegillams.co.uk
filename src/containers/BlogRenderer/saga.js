import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'utils/common/request';
import apiStructure from 'helpers/apiStructure';

import { selectState } from './selectors';
import { loadBlog, updateBlog } from './actions';

export function* doLoadBlog() {
  const currentState = yield select(selectState());
  const { blogId } = currentState;
  const requestURL = apiStructure.loadBlog.fullPath.split(':id').join(blogId);

  try {
    yield put(loadBlog.request());

    const result = yield call(request, requestURL, {
      method: 'GET',
    });

    if (result.error) {
      yield put(loadBlog.failure(result));
    } else {
      yield put(loadBlog.success(result));
    }
  } catch (err) {
    yield put(loadBlog.failure(err));
  } finally {
    yield put(loadBlog.fulfill());
  }
}

export function* doUpdateBlog() {
  const currentState = yield select(selectState());
  const { blogToUpdate } = currentState;
  const requestURL = apiStructure.updateBlog.fullPath;

  try {
    yield put(updateBlog.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(blogToUpdate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(updateBlog.failure(result));
    } else {
      yield put(updateBlog.success(result));
    }
    yield put(loadBlog.trigger(blogToUpdate.id));
  } catch (err) {
    yield put(updateBlog.failure(err));
  } finally {
    yield put(updateBlog.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadBlog.TRIGGER, doLoadBlog);
  yield takeLatest(updateBlog.TRIGGER, doUpdateBlog);
}
