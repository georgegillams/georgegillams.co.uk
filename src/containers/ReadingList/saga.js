import { takeLatest, put, call, select } from 'redux-saga/effects';

import request from 'client-utils/common/request';
import apiStructure from 'helpers/common/apiStructure';
import { selectState } from './selectors';

import { loadBooks, deleteBook, updateBook, createBook, setHighlightedBook } from './actions';

export function* doLoadBooks() {
  const requestURL = apiStructure.loadBooks.fullPath;
  const currentState = yield select(selectState());
  const { highlightedBook } = currentState;

  try {
    yield put(loadBooks.request());

    const result = yield call(request, requestURL, {
      method: 'GET',
    });

    if (result.error) {
      yield put(loadBooks.failure(result));
    } else {
      yield put(loadBooks.success(result));
    }
    if (highlightedBook) {
      setTimeout(() => {
        document.getElementById(highlightedBook)?.scrollIntoView?.({
          behavior: 'smooth',
        });
      }, 500);
    }
    yield put(setHighlightedBook.trigger(null));
  } catch (err) {
    yield put(loadBooks.failure(err));
  } finally {
    yield put(loadBooks.fulfill());
  }
}

export function* doCreateBook() {
  const currentState = yield select(selectState());
  const { bookToCreate } = currentState;
  const requestURL = apiStructure.createBook.fullPath;

  try {
    yield put(createBook.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(bookToCreate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(createBook.failure(result));
    } else {
      yield put(createBook.success(result));
      yield put(setHighlightedBook.trigger(result.id));
    }
    yield put(loadBooks.trigger());
  } catch (err) {
    yield put(createBook.failure(err));
  } finally {
    yield put(createBook.fulfill());
  }
}

export function* doUpdateBook() {
  const currentState = yield select(selectState());
  const { bookToUpdate } = currentState;
  const requestURL = apiStructure.updateBook.fullPath;

  try {
    yield put(updateBook.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify(bookToUpdate),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(updateBook.failure(result));
    } else {
      yield put(updateBook.success(result));
      yield put(setHighlightedBook.trigger(bookToUpdate.id));
    }
    yield put(loadBooks.trigger());
  } catch (err) {
    yield put(updateBook.failure(err));
  } finally {
    yield put(updateBook.fulfill());
  }
}

export function* doDeleteBook() {
  const currentState = yield select(selectState());
  const { bookToDelete } = currentState;
  const requestURL = apiStructure.deleteBook.fullPath;

  try {
    yield put(deleteBook.request());

    const result = yield call(request, requestURL, {
      method: 'POST',
      body: JSON.stringify({ id: bookToDelete }),
      headers: {
        'Content-Type': 'application/json',
      },
    });

    if (result.error) {
      yield put(deleteBook.failure(result));
    } else {
      yield put(deleteBook.success());
      yield put(setHighlightedBook.trigger(bookToDelete.id));
    }
    yield put(loadBooks.trigger());
  } catch (err) {
    yield put(deleteBook.failure(err));
  } finally {
    yield put(deleteBook.fulfill());
  }
}

export default function* saga() {
  yield takeLatest(loadBooks.TRIGGER, doLoadBooks);
  yield takeLatest(createBook.TRIGGER, doCreateBook);
  yield takeLatest(updateBook.TRIGGER, doUpdateBook);
  yield takeLatest(deleteBook.TRIGGER, doDeleteBook);
}
