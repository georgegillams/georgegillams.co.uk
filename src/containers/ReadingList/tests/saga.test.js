import { put, takeLatest } from 'redux-saga/effects';

import { loadBooks, deleteBook } from '../actions';

import saga, { doLoadBooks, doDeleteBook } from '../saga';

describe('BookList saga', () => {
  let mainSaga;

  beforeEach(() => {
    mainSaga = saga();
  });

  it('Should load books on loadBooks TRIGGER', () => {
    const takeLatestDescriptor = mainSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(loadBooks.TRIGGER, doLoadBooks));
  });

  describe('loadBooks actions', () => {
    let loadBooksGenerator;

    const response = {
      books: 'list of books',
      status: 200,
    };

    beforeEach(() => {
      loadBooksGenerator = doLoadBooks();

      const selectDescriptor = loadBooksGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call loadBooks.success on successful API call', () => {
      loadBooksGenerator.next({ highlightedBook: null });
      loadBooksGenerator.next();
      const putSuccess = loadBooksGenerator.next(response).value;
      loadBooksGenerator.next();

      expect(putSuccess).toEqual(put(loadBooks.success(response)));
    });

    it('Should call loadBooks.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      loadBooksGenerator.next({ highlightedBook: null });
      loadBooksGenerator.next();
      const putSuccess = loadBooksGenerator.next(response).value;
      loadBooksGenerator.next();

      expect(putSuccess).toEqual(put(loadBooks.failure(response)));
    });

    it('Should call loadBooks.failure if an exception occurs', () => {
      const response = new Error('Some error');
      loadBooksGenerator.next({ highlightedBook: null });
      const putFailure = loadBooksGenerator.throw(response).value;

      expect(putFailure).toEqual(put(loadBooks.failure(response)));
    });
  });

  describe('deleteBook actions', () => {
    let deleteBookGenerator;

    const response = {
      status: 200,
    };

    beforeEach(() => {
      deleteBookGenerator = doDeleteBook();

      const selectDescriptor = deleteBookGenerator.next().value;
      expect(selectDescriptor).toMatchSnapshot();
    });

    it('Should call deleteBook.success on successful API call', () => {
      deleteBookGenerator.next({ bookToDelete: 'some-book-id' });
      deleteBookGenerator.next();
      const putSuccess = deleteBookGenerator.next(response).value;
      deleteBookGenerator.next();

      expect(putSuccess).toEqual(put(deleteBook.success()));
    });

    it('Should call deleteBook.failure if API returns error', () => {
      const response = {
        error: 'error_type',
        errorMessage: 'Error message',
        status: 500,
      };
      deleteBookGenerator.next({ bookToDelete: 'some-book-id' });
      deleteBookGenerator.next();
      const putSuccess = deleteBookGenerator.next(response).value;
      deleteBookGenerator.next();

      expect(putSuccess).toEqual(put(deleteBook.failure(response)));
    });

    it('Should call deleteBook.failure if an exception occurs', () => {
      const response = new Error('Some error');
      deleteBookGenerator.next({ bookToDelete: 'some-book-id' });
      const putFailure = deleteBookGenerator.throw(response).value;

      expect(putFailure).toEqual(put(deleteBook.failure(response)));
    });
  });
});
