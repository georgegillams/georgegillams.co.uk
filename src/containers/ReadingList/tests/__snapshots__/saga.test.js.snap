// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookList saga deleteBook actions Should call deleteBook.failure if API returns error 1`] = `
{
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": {
    "args": [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BookList saga deleteBook actions Should call deleteBook.failure if an exception occurs 1`] = `
{
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": {
    "args": [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BookList saga deleteBook actions Should call deleteBook.success on successful API call 1`] = `
{
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": {
    "args": [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BookList saga loadBooks actions Should call loadBooks.failure if API returns error 1`] = `
{
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": {
    "args": [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BookList saga loadBooks actions Should call loadBooks.failure if an exception occurs 1`] = `
{
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": {
    "args": [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`BookList saga loadBooks actions Should call loadBooks.success on successful API call 1`] = `
{
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": {
    "args": [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
